### PR TITLE
[AEROGEAR-2568] Allow the DMZ URL to be set in the MAR view

### PR DESCRIPTION
**Pull request by JameelB**
## Description
Allow the DMZ Url for a mobile client to be set in the MAR view. This is added under the mobile client details section of the MAR view and should be able to be edited inline. Also add an `Edit YAML` option under the actions list in order to allow the DMZ Url to be edited within the YAML file. 

peek 2018-05-11 12-30

## Progress
- [x] Add a DMZ Url field under the `Mobile Client Details` section of the MAR view
- [x] Allow the DMZ Url to be edited inline
- [x] Add `Edit YAML` option under the actions list within the MAR view

## Verification Steps
**NOTE:** Requires changes from this [PR](https://github.com/openshift/origin-web-common/pull/320)

1. Provision a mobile application.
2. Go into the mobile client view
3. Verify that a field `DMZ Url` exists under the `Mobile Client Details` section.
4. Edit the DMZ url by clicking on the edit button
5. Enter an invalid url
6. Verify that an error message appears saying that the url provided is not valid and the save button is disabled.
7. Clear the input by clicking on the (x) button within the input field.
8. Set the DMZ url with a valid url
9. Verify that the DMZ url is successfully set. (The mobile client's `spec.dmzUrl` should be defined)
10. Verify that an `Edit YAML` option is available under the `Actions` list
11. Verify that the `spec.dmzUrl` property reflects what's shown in the `DMZ Url` field under the `Mobile Client Details`
12. Verify that no notification appears when the DMZ url is edited with no changes.

## Additional Notes
Related JIRA Ticket - https://issues.jboss.org/browse/AEROGEAR-2568

### CODE REVIEW

File: app/scripts/controllers/mobileClients.js [hunk 1]
```diff
@@ -32,6 +33,39 @@ angular.module('openshiftConsole')
           title: $routeParams.mobileclient
         }
       ];
+      ctrl.editDmzUrl = false;
+
+      ctrl.setDmzUrl = function(dmzUrl) {
+        var mobileClient = angular.copy(ctrl.mobileClient);
+        if (dmzUrl && dmzUrl != "") {
```
Comment (dimitraz) [inline app/scripts/controllers/mobileClients.js]: not sure if they prefer the lodash version of this i.e. `_.isEmpty()`

File: app/scripts/controllers/mobileClients.js [hunk 2]
```diff
@@ -32,6 +33,39 @@ angular.module('openshiftConsole')
...
+          mobileClient.spec['dmzUrl'] = dmzUrl;
+        } else {
+          delete mobileClient.spec['dmzUrl'];
```
Comment (dimitraz) [inline app/scripts/controllers/mobileClients.js]: again not sure if they prefer the lodash route, i think `_.unset()` is similar?

File: app/scripts/controllers/mobileClients.js (hunk 1 above)
Comment (JameelB) [reply app/scripts/controllers/mobileClients.js]: Was actually supposed to remove that. Good catch :+1:

**Comment by sedroche** [comment]
The workflow looks good @JameelB.
I see that Patternfly have an inline edit as well -> http://www.patternfly.org/pattern-library/forms-and-controls/inline-edit/#code

inline-edit

I would recommend trying to re-use their html and styling where possible (The component is not available for angular, just react, but it will suit for moving to react in the future)

Also, from testing I notice we get a notification for update even if the value is empty or no change happens. I would only notify if a change happens.

**Comment by philbrookes** [comment]
Verified locally.

File: app/scripts/directives/inlineEdit.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    patternErrmsg: '<',
+    value: '<',
+    saveFunc: '<'
+  },
+  templateUrl: 'views/directives/inline-edit.html'
+});
+
+function InlineEdit() {
+  var ctrl = this;
+
```
Comment (philbrookes) [inline app/scripts/directives/inlineEdit.js]: Should we add a `ctrl.saveFunc = ctrl.saveFunc || function(){};`, so that line 38 won't break if no function is passed in?

File: app/scripts/directives/inlineEdit.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  controller: [
+    InlineEdit
+  ],
+  controllerAs: '$ctrl',
+  bindings: {
+    label: '<',
+    pattern: '<',
+    patternErrmsg: '<',
+    value: '<',
+    saveFunc: '<'
```
Comment (sedroche) [inline app/scripts/directives/inlineEdit.js]: Does `onEdited` make more sense? Component doesn't know what happens to the value on edit.

File: app/scripts/controllers/mobileClients.js [hunk 5]
```diff
@@ -33,6 +34,30 @@ angular.module('openshiftConsole')
...
       ];
+      ctrl.validUrlPattern = VALID_URL_PATTERN;
+
+      ctrl.setDmzUrl = function(dmzUrl) {
+        var mobileClient = angular.copy(ctrl.mobileClient);
+        if (!dmzUrl) {
+          dmzUrl = undefined;
+        }
+
+        if (mobileClient.spec['dmzUrl'] !== dmzUrl) {
```
Comment (sedroche) [inline app/scripts/controllers/mobileClients.js]: Can we move this check to `inlinedit`? `saveFunc` should only be triggered when a change happens

File: app/styles/_inline-edit.less [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+.form-control-pf-value:hover {
+  background-color: transparent;
+  border: 1px solid lightgray;
+  padding-left: 0px;
+}
+
+.form-control-pf-textbox {
+  display: inline-block;
+  position: relative;
```
Comment (sedroche) [inline app/styles/_inline-edit.less]: css should be alphabetical order
